cpinstaller/src/cpinstaller.js

```diff
@@ -391,466 +792,371 @@ export class CPInstallButton extends InstallButton {
-            await disconnect();
-            updateUIConnected(this.connectionStates.DISCONNECTED);
-        });
-    });
-    //butClear.addEventListener("click", clickClear);
-    butProgram.addEventListener("click", clickProgram);
-    butProgramBootloader.addEventListener("click", clickProgramNvm);
-    for (let i = 0; i < partitionData.length; i++) {
-        partitionData[i].addEventListener("change", checkProgrammable);
-        partitionData[i].addEventListener("keydown", checkProgrammable);
-        partitionData[i].addEventListener("input", checkProgrammable);
-    }
-    //autoscroll.addEventListener("click", clickAutoscroll);
-    //baudRate.addEventListener("change", changeBaudRate);
 
-    // handle runaway errors
-    window.addEventListener("error", event => {
-        console.warn(`Uncaught error: ${event.error}`);
-    });
+    //////////////// OTHER HELPERS ////////////////
 
-    // handle runaway rejections
-    window.addEventListener("unhandledrejection", event => {
-        console.warn(`Unhandled rejection: ${event.reason}`);
-    });
+    async saveCredentials() {
+        this.saveSetting('CIRCUITPY_WIFI_SSID');
+        this.saveSetting('CIRCUITPY_WIFI_PASSWORD');
+        this.saveSetting('CIRCUITPY_WEB_API_PASSWORD');
+        this.saveSetting('CIRCUITPY_WEB_API_PORT');
 
-    // WebSerial feature detection
-    if ("serial" in navigator) {
-        const notSupported = document.getElementById("notSupported");
-        notSupported.classList.add("hidden");
+        await this.writeSettings(this.tomlSettings);
+        if (this.hasNativeUsb()) {
+            //this.setBootDisabled(true);
+        }
+        await this.nextStep();
     }
 
-    //initBinSelector();
-    //initBaudRate();
-    loadAllSettings();
-    this.logMsg("CircuitPython ESP32 Installer loaded.");
-    checkProgrammable();
-});
-
-let latestFirmwares = []
-
+    getSetting(setting, defaultValue = '') {
+        if (this.tomlSettings && this.tomlSettings.hasOwnProperty(setting)) {
+            return this.tomlSettings[setting];
+        }
 
-function doThingOnClass(method, thing, classSelector) {
-    const classItems = document.getElementsByClassName(classSelector)
-    for (let idx = 0; idx < classItems.length; idx++) {
-        classItems.item(idx).classList[method](thing)
+        return defaultValue;
     }
-}
 
-function setDefaultBoard() {
-    const board = getFromQuerystring(QUERYSTRING_BOARD_KEY)
-    if (board && hasBoard(board)) {
-        binSelector.value = board
-        return true
-    }
-}
+    async getBootDisabled() {
+        // This is a very simple check for now. If there is something more complicated like a disable
+        // command behind an if statement, this will not detect it is enabled.
+        let fileContents;
+        if (this.repl) {
+            return true; // Always disabled in this case
+        } else if (this.circuitpyDriveHandle) {
+            fileContents = await this.readFile("boot.py");
```

Comment:
Can you hit this case? I don't think you can have a handle to the CIRCUITPY drive if it is disabled.

Reply:
When the UF2 is first installed, the CIRCUITPY drive is enabled, which makes writing through the REPL impossible. So for boards with the CIRCUITPY drive it currently needs to take this route until it is able to be disabled (a future feature).

Reply:
One of the reasons for this is maybe somebody wants to use the installer, but doesn't want to use the Web Workflow to edit files and feels more comfortable using the CIRCUITPY drive.

I should also add, one of my goals was to keep user interaction as minimal as possible, so if they are doing web workflow, I was thinking of disabling CIRCUITPY at the last moment. But upon further thinking, it may be better to do that right away and then use the REPL so that we can get the IP address. For now, I'm deferring the decision.